[JENKINS-54287] GSoC student guideline update (#1905)
jenkins-infra/jenkins.io: content/projects/gsoc/students.adoc

== Student Application process

> === Eligibility steps

Reviewer: 👍

You must verify that you are eligible to participate in the program.
We cannot make any exception. Please:

. Make sure you are link:https://developers.google.com/open-source/gsoc/faq#what_are_the_eligibility_requirements_for_participation[eligible to participate in the program]
. If you are an accepted students in the United States of America on an F1 visa, you must obtain an link:https://developers.google.com/open-source/gsoc/faq#i_am_an_accepted_student_in_the_united_states_on_an_f1_visa_how_do_i_get_authorization_to_participate[authorization before you can participate].

=== Application steps

. Check out the link:/projects/gsoc/2019/project-ideas[GSoC 2019 Project ideas]
. Select an interesting project idea or draft your own proposal.
. If you are not familiar with Jenkins, read the introductory info on the website and try it out with one of your previous projects.

> . Join the _jenkinsci-dev_ and _jenkinsci-gsoc-all-public_ mailing lists, add GSoC office hours to your calendar

Reviewer: @martinda I suggest you also add:
> Join https://gitter.im/jenkinsci/gsoc-sig

Author: I added it to the community bonding period, after talking to mentors.

 * Say something when you are stuck, lost in the code, confused about the objectives, etc.
 * Produce good quality code with reasonable amount of testing and documentation.
 * Have a finalized deliverable at the end of the project.
 ** For plugin development projects, this means releasing a plugin to the alpha or to the official update center.
. Take Time off
 * You have approximately 5 "vacation days" during the project, do not hesitate to use them if required.
 * Notify your mentors in advance when you take time off.
 * Use weekends to have a rest, avoid significant overwork and enjoy coding
 * Timely notify mentors in the case of emergencies and outages (missing scheduled meetings, etc.).
 * Timely notify mentors and org admins about unexpected time commitments (life goes on, it is normal - mentors will let you know if they can't be reached too).
. Be present on-line
 * Be around in _#jenkins_ IRC and in the project chats during the working hours (the link:https://gitter.im/jenkinsci/gsoc-sig[Jenkins GSoC Gitter Chat], and the Gitter Chat of your project)
 * Be proactive; reach out to the community if required

>  * Optional: Attend Jenkins governance meetings if the timezone allows

Reviewer: @martinda _Just a thought: Perhaps action items can be split into a checklist format?_

**Required _checklist_**
☑Be in Jenkins GSoC Gitter Chat
☐ Be in Gitter Chat of your project
☐ Be proactive; reach out to the community if required

**Optional _checklist_**
☐ Be around in _#jenkins_ IRC
☐ Attend Jenkins governance meetings if the timezone allows

**Not expected _checklist_**
☒ Strictly follow the originally submitted mini-design and the project proposal
☒ Investigate and solve *every* issue on your own

_( a student can print this document, and mark completed items )_

Students are **not** expected to...

. Strictly follow the originally submitted mini-design and the project proposal
 * The world is not ideal, and there may be unexpected obstacles or shortcuts
 * Upon the discussion with mentors, any plan can be adjusted
 * We expect students to achieve at least some goals in the original proposal

> . Investigate and solve *every* issue on their own

Reviewer: suggestion: change "their own" to "your own"

It will help you to polish the proposal,
and mentors will consider contributions and interactions with the community when processing applications.

[[UsefulLinks]]
=== Useful links

Here are a list of links to help you get started on participating in Jenkins and in coding for the Jenkins project,
in increasing level of difficulty.

> * link:https://jenkins.io/project/governance/[Project Governance]

Reviewer 2: Links should be relative, e.g. `/project/governance/`

Author: Fixed this one and all similar ones, will push momentarily.

> * link:https://github.com/jenkinsci/jenkins/blob/master/CONTRIBUTING.md[Contributing to Jenkins] (for contributions to the core)

Reviewer 2: Not sure it is the best link to start from

Author: Do you have another link to suggest? I moved it below a different title.

> * link:https://wiki.jenkins.io/display/JENKINS/Beginners+Guide+to+Contributing[Beginners Guide to Contributing]

Reviewer 2: Replace by https://jenkins.io/participate/ ? I think many links can be replaced by this one

* link:https://wiki.jenkins.io/display/JENKINS/Plugin+tutorial[Plugin tutorial]

> * link:https://jenkins.io/doc/developer/tutorial/[Another plugin tutorial]

Reviewer 2: Blogpost link is enough, old tutorials are not needed imho

* link:https://jenkins.io/blog/2017/08/07/intro-to-plugin-development/[Plugin Development Tutorials, Videos, and More]
* link:https://jenkins.io/doc/developer/plugin-development/pipeline-integration/[Writing Pipeline-Compatible Plugins]
* link:https://jenkins.io/doc/developer/extensions/[Extension Points]

> * link:https://wiki.jenkins.io/display/JENKINS/Hosting+Plugins[Releasing plugins]

Reviewer 2: Not needed for beginners

Author: Moved under an advanced title.

Reviewer 2: Maybe linking to the developer guidelines would be enough. I do not want to overload the page

There is also a
link:https://issues.jenkins-ci.org/issues/?jql=labels%20%3D%20newbie-friendly[list of newbie-friendly issues].

Feel free to contact potential mentors and org admins if you need help with choosing a newbie friendly issue to tackle.

> Having a look at recent plugins can help you getting familiar with the coding style and the type of code we write:

Reviewer 2: Is it really needed? It references only Java examples, and not all links here are the codestyles I would recommend

* link:https://github.com/jenkinsci/remoting-kafka-plugin[Remoting over Kafka plugin]
* link:https://github.com/jenkinsci/external-workspace-manager-plugin/[External Workspace Manager plugin]

== Student Selection process

Once the application period is over, administrators and mentors make a decision on which proposal to accept
based on the proposal submitted to the Google Summer of Code website.
Only proposals submitted before the deadline to the Google Summer of Code website are considered.

We understand students are anxious to know whether they are selected or not, but admins and mentors are
bound to secrecy until Google announces the selection results.
We will not discuss the selection with students until Google makes the announcement.

== Congratulations, you have been selected... or not

We thank all students who reach out to us during the application period.
If you have not been selected link:https://google.github.io/gsocguides/student/being-turned-down[read this],
there could be many reasons, and some are even outside of our control.
Do not feel bad, we encourage you to stay with the community, and apply again next year.

> If you have been selected, the community bonding period starts immediately.

Reviewer 2: Unlikely. The student needs some time to prepare. I would say "within 2 days after the announcement" or so, mentors are expected to reach out within this time

Reviewer 3: ```suggestion
If you have been selected, the community bonding period starts within 2 days after the announcement.
```

It will help you to polish the proposal,
and mentors will consider contributions and interactions with the community when processing applications.

[[UsefulLinks]]
=== Useful links

Here are a list of links to help you get started on participating in Jenkins and in coding for the Jenkins project,
in increasing level of difficulty.

* link:/project/governance/[Project Governance]
* link:/participate[Participate and Contribute]
* link:https://wiki.jenkins.io/display/JENKINS/Plugin+tutorial[Plugin tutorial]
* link:/blog/2017/08/07/intro-to-plugin-development/[Plugin Development Tutorials, Videos, and More]

> Some older tutorials (warning, some may be a bit out of date):

Reviewer 2: Linked from the blogpost above, maybe an overkill

Author: I removed this mini-section.

* link:/doc/developer/tutorial/[Another plugin tutorial]
* link:/doc/developer/plugin-development/pipeline-integration/[Writing Pipeline-Compatible Plugins]
* link:/doc/developer/extensions/[Extension Points]

For advanced level, when you get ready to release a plugin project:

* link:https://wiki.jenkins.io/display/JENKINS/Hosting+Plugins[Releasing plugins]

For contributing to the core of Jenkins (as opposed to a plugin contribution):

> * link:https://github.com/jenkinsci/jenkins/blob/master/CONTRIBUTING.md[Contributing to Jenkins]

Reviewer 2: If we remove the two lists above, could be the fifth bullet in the first list (after plugin links)

Author: done

There is also a
link:https://issues.jenkins-ci.org/issues/?jql=labels%20%3D%20newbie-friendly[list of newbie-friendly issues].

Feel free to contact potential mentors and org admins if you need help with choosing a newbie friendly issue to tackle.

> 

Reviewer 2: ```suggestion
See the contact links in project proposals.

```

== Student Selection process

Once the application period is over, administrators and mentors make a decision on which proposal to accept
based on the proposal submitted to the Google Summer of Code website.
Only proposals submitted before the deadline to the Google Summer of Code website are considered.

We understand students are anxious to know whether they are selected or not, but admins and mentors are
bound to secrecy until Google announces the selection results.
We will not discuss the selection with students until Google makes the announcement.

== Congratulations, you have been selected... or not

We thank all students who reach out to us during the application period.
If you have not been selected link:https://google.github.io/gsocguides/student/being-turned-down[read this],
there could be many reasons, and some are even outside of our control.
Do not feel bad, we encourage you to stay with the community, and apply again next year.

> If you have been selected, the community bonding period starts within two days afterthe announcement.

Reviewer 2: ```suggestion
If you have been selected, the community bonding period starts within two days after the announcement.
```

== Community Bonding

As soon as the students are accepted, the community bounding period starts.
During this period, students are not expected to be coding immediately.
Instead they are expected to prepare to code.

A successful community bonding usually leads to successful coding periods.
It is our experience that poor community bonding leads to difficult coding periods.

Use the community bonding to:

> . Formalize the communication channels with your mentors:

Reviewer 2: ```suggestion
. Define the communication channels with your mentors:
```

>   * Setup a gitter chat room for the project (linked to github).

Reviewer 2: ```suggestion
  * If it does not exist, setup a Gitter chat room for your project project.
```

. Setup the weekly meeting schedule with your mentors:
  * Two meetings per week is recommended,
  * Announce your meeting schedule to:
  ** The link:https://groups.google.com/forum/#!forum/jenkinsci-gsoc-all-public[Jenkins GSoC mailing list].
  ** The gitter chat of your project.
  ** Send a google calendar meeting invite to the mentors, CC the org admins.

> . Get introduced to the key stakeholders and contributors in the area of the project by you mentors:

Reviewer 2: ```suggestion
. Get introduced to the key stakeholders and contributors in the area of the project by your mentors:
```

Author: Got it.

  * For example, an introduction to subject matter experts.
. Continue to discuss and plan the project with the community and the mentors:
  * Work on the design document of the project.
  ** Work on clarifying objectives and expectations,
  ** Study, refine and discuss the design and the project plan,
  ** Top-level architecture document:
  *** Create diagrams of operation,
  *** Answer questions such as "How is the user going to use this?", "What configurations are needed?", etc.,
  *** Some people find it useful to write a mini user guide or how-to guide, as if the project was already done. This usually helps define the project.,
  * Create an implementation plan with milestones per coding period.

>   * At this point it may be appropriate to discuss the project on the _jenkinsci-dev@googlegroups.com_ mailing list (link:https://groups.google.com/forum/#!forum/jenkinsci-dev[archives]). Talk to the mentor about it.

Reviewer 2: ```suggestion
  * At this point it may be appropriate to discuss the project on the _jenkinsci-dev@googlegroups.com_ mailing list or on the SIG mailing list. Talk to your mentors about it.
```

Author: Got it.

. Setup your computer and your development environment to work on the project (see <<UsefulLinks>>).
. Learn and discuss the process with the mentors:
  * Setup the github project,
  * Pull-requests,
  * Code reviews.

> . Learn the coding style expectations:

Reviewer 2: It is too detailed IMHO. Maybe you could put the list in the beginning of the Wiki page and just keep a link here

Author: @oleg-nenashev  Not sure where you want the link to be, and where you want the content of the list to be.

  * From the link:https://wiki.jenkins.io/display/JENKINS/Code+Style+Guidelines[Code Style Guidelines]
  * By reading existing code

>   * When in doubt, ask mentors or submit a pull-request and ask for review

Reviewer 2: Unrelated to code style?

Author: Got it. I will remove when I move the list elsewhere.

Reviewer 2: What about creating a "Code Style Best Practices" section in the doc for now?
And then this list can just refer them

Author: created such section and pushed my commit.

  * The following are important for functionality and quality:
  ** Proper annotations,
  ** Unit and integration testing in Jenkins.
  * If your project is a plugin, you also need to learn the link:https://wiki.jenkins.io/display/JENKINS/Hosting+Plugins[plugin release process].
. We use link:https://issues.jenkins-ci.org/secure/Dashboard.jspa[Jira] to track GSoC tasks:
  * Create an account using link:https://accounts.jenkins.io/login[this link].
  * Become familiar with navigating Jira.

== Coding periods

Students are expected to...

. Work on the GSoC project as it is a full-time employment.
 * It means that 30..40 hours per week is an **expected** workload though it can be adjusted upon the agreements with mentors.
 * Push code to github almost every day of every coding period.
 * Chat a line or two about what you are doing, almost every coding day, in your project channel (writing code, writing tests, updating documentation, etc.).
 ** Just saying "Hi, today I am working on these classes" or "writing tests for ..." is good enough, but you can of course interact more as needed.
 * Write a short summary of the work done each week, published to:
 ** A personal blog, or

>  ** To the Jenkins website blogs (see link:https://github.com/jenkins-infra/jenkins.io/blob/master/CONTRIBUTING.adoc#adding-a-blog-post[adding a blog post])

Reviewer 2: or to the SIG mailing list.

Reviewer 2: I also do not think that blogposting to jenkins.io each week is a good idea. Posting for milestones should be enough

Author: I agree.

 * Produce good quality code with reasonable amount of testing and documentation.
 * Have a finalized deliverable at the end of the project.
 ** For plugin development projects, this means releasing a plugin to the alpha or to the official update center.
 ** Have documentation on how to use the plugin of the features developed during the project.
 *** Documentation usually starts at the README file of the github repository
 *** The format is either link:https://guides.github.com/features/mastering-markdown/[Github Markdown] or link:https://asciidoctor.org/docs/[Asciidoctor].
. Take Time off
 * You have approximately 5 "vacation days" during the project, do not hesitate to use them if required.
 * Notify your mentors in advance when you take time off.
 * Use weekends to have a rest, avoid significant overwork and enjoy coding
 * Timely notify mentors in the case of emergencies and outages (missing scheduled meetings, etc.).
 * Timely notify mentors and org admins about unexpected time commitments (life goes on, it is normal - mentors will let you know if they can't be reached too).
. Be present on-line

>  * Be around in _#jenkins_ IRC and in the project chats during the working hours (the link:https://gitter.im/jenkinsci/gsoc-sig[Jenkins GSoC Gitter Chat], and the Gitter Chat of your project)

Reviewer 2: ```suggestion
 * Be around in the project chats during the working hours (the link:https://gitter.im/jenkinsci/gsoc-sig[Jenkins GSoC Gitter Chat], and the Gitter Chat of your project)
```

* GSoC 2016 blog post:
** link:/node/tags/external-workspace-manager/[External Workspace Manager]

== Getting in touch

=== Mailing lists

Since the Jenkins community is distributed across all time zones,
and since the gitter chat rooms are more difficult to search,
we recommend using mailing lists for the most of communications.

Students must join the Jenkins GSoC mailing list:

* _jenkinsci-gsoc-all-public@googlegroups.com_ - sync-ups on organizational topics (meeting scheduling, process Q&A) (link:https://groups.google.com/forum/#!forum/jenkinsci-gsoc-all-public[archives]).

After talking to the org admins and/or the project mentors, and once the project is ready to be discussed with the developers,
the student should join the developer mailing list:

* _jenkinsci-dev@googlegroups.com_ - for all technical discussions and the project application (link:https://groups.google.com/forum/#!forum/jenkinsci-dev[archives]).

Organizational questions:

* _jenkins-gsoc-2019-org-admins@googlegroups.com_ - for **private** communications with org admins (escalations, issues with mentors)
** Please *DO NOT* use this mailing list for applications and intro emails

=== Chat

> We use the link:/chat[#jenkins_ IRC channel] and the link:https://gitter.im/jenkinsci/gsoc-sig[Jenkins GSoC Gitter Chat]

Reviewer 2: ```suggestion
We use the link:https://gitter.im/jenkinsci/gsoc-sig[Jenkins GSoC Gitter Chat]
```

. Work on the GSoC project as it is a full-time employment.
 * It means that 30..40 hours per week is an **expected** workload though it can be adjusted upon the agreements with mentors.
 * Push code to github almost every day of every coding period.
 * Chat a line or two about what you are doing every coding day in your project channel (writing code, writing tests, updating documentation, etc.).
 * You have timely community interactions when you need help (do not stay stuck without telling mentors).
 * If you are stuck, lost in the code, confused, do say it.
 * Produce the good quality code with reasonable amount of testing and documentation.
 * Have a finalized deliverable at the end of the project.
. Take Time off
 * You have approximately 5 "vacation days" during the project, do not hesitate to use them if required.

>  * Notify your mentors when you take time off.

Author: Fixed.

 * Use weekend to have a rest, avoid significant overwork and enjoy coding
 * Timely notify mentors in the case of emergencies and outages (missing scheduled meetings, etc.).
 * Timely notify mentors and org admins about unexpected time commitments
. Be present on-line

>  * Be around in _#jenkins_ IRC and in the project chats during the working hours

Author: Fixed.

>  * Attend Jenkins governance meetings if the timezone allows

Author: Made optional.

=== Evaluations

At the end of each coding period, students are expected to do a public on-line presentation.

> . The presentation consists of Google Slides and a demo, on a Google Hangout.

Author: Fixed.

See the link:/projects/gsoc/#office-hours[main GSoC page] for the schedule.

== Post-GSoC

Congratulations, you have made it to the end!

Once GSoC is over, final results are announced by Google. But this is not the end of the road.

You can:

. Continue to develop your project within the Jenkins community
. Participate in other Jenkins projects
. Participate again next year
. Become a mentor in link:https://summerofcode.withgoogle.com[Google Summer of Code] for next year
. Become a mentor in link:https://codein.withgoogle.com/[Google Code In]

Depending on the project results, and available budget, we may offer a sponsored trip
to link:https://www.cloudbees.com/devops-world/[DevOps World - Jenkins World] or another Jenkins-related event to students

> who successfully finish your projects.

Reviewer 2: their

Reviewer 3: ```suggestion
who successfully finish their projects.
```

See the link:/projects/gsoc/#office-hours[main GSoC page] for the schedule.

== Post-GSoC

> Congratulations, you have made it to the end!

Reviewer 2: Should be an easter egg here :)

Author: Er... I know what an easter egg is but I am not sure if you want me to do something here.

Once GSoC is over, final results are announced by Google. But this is not the end of the road.

You can:

> . Continue to develop your project within the Jenkins community

Reviewer 2: Add: present your project at a local Jenkins Area Meetup (https://jenkins.io/projects/jam/ , no header anchors unfortunately)

Author: done

. Participate in other Jenkins projects
. Participate again next year
. Become a mentor in link:https://summerofcode.withgoogle.com[Google Summer of Code] for next year
. Become a mentor in link:https://codein.withgoogle.com/[Google Code In]

Depending on the project results, and available budget, we may offer a sponsored trip
to link:https://www.cloudbees.com/devops-world/[DevOps World - Jenkins World] or another Jenkins-related event to students
who successfully finish their projects.
This sponsorship is not guaranteed though.

> If students agree to go to such event, we expect students to present your project and to write a blog-post about the trip.

Reviewer 2: Add link to the blogpost from @pvtuan10 ?

Author: got it
